tests/test_storage.py

```diff
@@ -40,25 +40,41 @@ def test_metadata_target(tmp_metadata_target):
 
 
 @pytest.mark.parametrize(
-    "file_paths", [lazy_fixture("netcdf_local_paths"), lazy_fixture("netcdf_http_paths")]
+    "file_paths",
+    [
+        lazy_fixture("netcdf_local_paths"),
+        lazy_fixture("netcdf_http_paths"),
+        lazy_fixture("netcdf_http_paths_with_secrets"),
+    ],
 )
 @pytest.mark.parametrize("copy_to_local", [False, True])
 @pytest.mark.parametrize("use_cache, cache_first", [(False, False), (True, False), (True, True)])
 @pytest.mark.parametrize("use_dask", [True, False])
 @pytest.mark.parametrize("use_xarray", [True, False])
+@pytest.mark.parametrize("use_query_string_secrets", [True, False])
 def test_file_opener(
-    file_paths, tmp_cache, copy_to_local, use_cache, cache_first, dask_cluster, use_dask, use_xarray
+    file_paths,
+    tmp_cache,
+    fake_secrets,
+    copy_to_local,
+    use_cache,
+    cache_first,
+    dask_cluster,
+    use_dask,
+    use_xarray,
+    use_query_string_secrets,
 ):
     all_paths, _ = file_paths
     path = str(all_paths[0])
     cache = tmp_cache if use_cache else None
+    secrets = fake_secrets if use_query_string_secrets and file_paths[0][-1] == "?" else None
 
     def do_actual_test():
         if cache_first:
-            cache.cache_file(path)
+            cache.cache_file(path, secrets)
```

Comment:
Good point. I think I've resolved this with https://github.com/pangeo-forge/pangeo-forge-recipes/commit/c42151e6ddce7123e4349748ee54b059f9fb8fb0

Reply:
Yes, now I think secrets will be populated when we want it populated.

However, I'm still not convinced we are really testing that the _correct_ URL is being generated. the base test fixture `netcdf_http_paths` creates an http server with files at the specified URLs. This server doesn't care about the query string at all--it basically ignores it. So we could pass any `secrets`, or no `secrets`, and this test would still pass. So that means we can't count on this test to ensure that the secrets feature is working correctly. We could break it in the future and the test would not notify us.

The right thing to do is probably to refactor [http_auth_server.py](https://github.com/pangeo-forge/pangeo-forge-recipes/blob/master/tests/http_auth_server.py) to add an additional option for a query string secret. You would have to
1. Refactor the command line arguments to allow for keyword (rather than position-based) arguments here 
   https://github.com/pangeo-forge/pangeo-forge-recipes/blob/c4867719c4866ec46436b5d53d95ceb470a6f72e/tests/http_auth_server.py#L8-L11
   [Clic](https://click.palletsprojects.com/en/8.0.x/) would probably be the best choice.
2. Add a new option, e.g. `--required-query-string=foobar`
3. Refactor the `netcdf_http_paths` test a bit to work with this new syntax
4. While you're at it, you could also add a test for standard username / password, which is currently possible but not actually implemented in a test.